Added type hints for Pytorch Marian calls

clefourrier opened the pull request:
As the title suggests, this PR adds type hints to the PyTorch Marian calls. (#16059 )

@Rocketknight1

Commit:
Added type hinting for forward functions in pytorch marian

Commit:
typo correction

patil-suraj commented on src/transformers/models/marian/modeling_marian.py:
```diff
@@ -76,7 +76,7 @@ def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int, decoder_start
 
 # Copied from transformers.models.bart.modeling_bart._make_causal_mask
-def _make_causal_mask(input_ids_shape: torch.Size, dtype: torch.dtype, past_key_values_length: int = 0):
+def _make_causal_mask(input_ids_shape: torch.Size, dtype: torch.dtype, past_key_values_length: int = 0) -> torch.Tensor:
```
same comment as above.

clefourrier replied:
same

clefourrier commented on src/transformers/models/marian/modeling_marian.py:
```diff
@@ -59,7 +59,7 @@
 
 # Copied from transformers.models.bart.modeling_bart.shift_tokens_right
-def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int, decoder_start_token_id: int):
+def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int, decoder_start_token_id: int) -> torch.Tensor:
```
I will remove it from this PR then

clefourrier commented on src/transformers/models/marian/modeling_marian.py:
```diff
@@ -92,7 +92,7 @@ def _make_causal_mask(input_ids_shape: torch.Size, dtype: torch.dtype, past_key_
 
 # Copied from transformers.models.bart.modeling_bart._expand_mask
-def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None) -> torch.Tensor:
```
same

Commit:
Removed type hints on functions from BART per Suraj Patil request

Commit:
fix import pb

Commit:
fix typo

Commit:
corrected tuple call

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_16200). All of your documentation changes will be reflected on that endpoint.

Commit:
ran black

clefourrier commented:
Hm, do I need to change the BART too for consistency?  
```
src/transformers/models/marian/modeling_marian.py: copy does not match models.bart.modeling_bart.BartAttention at line 144
src/transformers/models/marian/modeling_marian.py: copy does not match models.bart.modeling_bart.BartEncoderLayer at line 289
src/transformers/models/marian/modeling_marian.py: copy does not match models.bart.modeling_bart.BartDecoderLayer at line 356
src/transformers/models/marian/modeling_marian.py: copy does not match models.bart.modeling_bart.BartForCausalLM at line 1529
src/transformers/models/pegasus/modeling_pegasus.py: copy does not match models.marian.modeling_marian.MarianSinusoidalPositionalEmbedding at line 109
src/transformers/models/roformer/modeling_roformer.py: copy does not match models.marian.modeling_marian.MarianSinusoidalPositionalEmbedding at line 73
Run `make fix-copies` or `python utils/check_copies.py --fix_and_overwrite` to fix them.
```

Running `make fix-copies` basically reverts all my changes.

Rocketknight1 commented:
Hi @clefourrier, sorry for the delay! We will indeed need to make changes to BART to make the tests work - but we already have an existing BART PR here: https://github.com/huggingface/transformers/pull/16270. Probably the easiest thing to do is:

1. Wait until we merge that PR
2. If any changes in Marian are still needed, rebase your current PR onto it (make sure to force push afterwards!)
3. Check tests again afterwards

clefourrier commented:
Hi @Rocketknight1 , perfect! Can you ping me once the BART PR is merged then?

Rocketknight1 commented:
@clefourrier It was just merged about half an hour ago! You'll probably need to pull changes from upstream to the master branch of your fork, and then rebase (with force push!) onto that.

Commit:
after fix-copies
Some optional tags on primitives were removed, past_key_values in MarianForCausalLM changed from Tuple of Tuple to List

clefourrier commented:
@Rocketknight1 Well, it seems the last failures comes from children models (pegasus and roformer) - do they have their own PR?

Rocketknight1 committed:
Fixing copies to roformer and pegasus

Rocketknight1 commented on src/transformers/models/marian/modeling_marian.py:
```diff
@@ -477,7 +477,7 @@ class MarianPreTrainedModel(PreTrainedModel):
     supports_gradient_checkpointing = True
 
-    def _init_weights(self, module):
+    def _init_weights(self, module: Union[nn.Linear, nn.Embedding, MarianSinusoidalPositionalEmbedding]):
```
A lesser contributor would have just given up and written `Any` for this one. I respect it.

Rocketknight1 reviewed (approved):
PR looks great! The issue with Pegasus and Roformer was just that they copied the sinusoidal embeddings from `Marian`. Running `make fix-copies` after the rebase fixed the issue, so this should be ready to merge once the tests are green!
